numpy/core/src/multiarray/arrayobject.c

```diff
@@ -1086,7 +1086,8 @@ _strings_richcompare(PyArrayObject *self, PyArrayObject *other, int cmp_op,
     int val;
 
     /* Cast arrays to a common type */
```

Comment:
That whole casting if, could probably be moved to the new iterator, but I think that is good enough on another day maybe.